---
title: "Cross validation for the trials- for looping for factor with marker and spatial
  analysis"
author: "Tesfahun A. Setotaw"
date: "2023-03-17"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## Introduction

This script is desinged to make cross validation analysis of the trials with marker and spatial information.

```{r cars}
# Load the packages necessary for the analysis
library(tidyverse)
library(sommer)
library(caret)
library(dplyr)
#########################################
# Import phenotype data 
#######################################
setwd("/Users/tas286/Documents/Data_geno_pheno_for_selected_trials")
phen = read.csv("Phenotype_YldQt_Val_2014.csv")
head(phen)
## Change the factor variable to factor
set.seed(123)
phen$geno = as.factor(phen$germplasmName)
phen$rowf = as.factor(phen$rowNumber)
phen$colf = as.factor(phen$colNumber)
phen$col = phen$colNumber
phen$row = phen$rowNumber
phen$loc = as.factor(phen$locationName)
str(phen)
colnames(phen)[33] = "Grain_protein"
colnames(phen)[34] = "Test_weight"
colnames(phen)[35] = "Grain_Yield"

Traits = c("Grain_protein","Test_weight", "Grain_Yield" )

##############################################
# Read the SNPs data of the genotypes from VCF format and change to dosage
##############################################

library(vcfR) # load the vcfR package
snps = read.vcfR(file = "Imputed_data/Imputed/YldQT_Val_2014_Imputed.vcf")
head(snps[,1:20])

snps_num <- vcfR::extract.gt(snps,
                             element = "GT",
                             IDtoRowNames  = T, # to list the row name or not
                             as.numeric = T,
                             convertNA = T
)


snps <- t(snps_num)

summary(cars)
```

## Spliting the data as training and testing set and Analysis

In this section the data will be divided as training and testing set for the 5 fold cross validation. 

```{r pressure, echo=FALSE}
########################################33333#############
# Grouping the training and testing set for 5 fold validation
#  Preparing tibbles for output storage

# the tibble to store the predictability output from the model
pr_y = tibble()

# The tibble to store the Root Mean Squared Error (RMSQ)  and Mean Absolute Error (MAD)
Accu = tibble()

A1 = A.mat(snps) # The addative relationship matrix
Env = levels(phen$loc)


for(env in Env){
  SL <- subset(x = phen, subset = loc == env) # Subseting the data for each env
  TraitN = colnames(SL[Traits])[colSums(is.na(SL[Traits])) < 25] # selecting the trait without NA
  ntt = length((TraitN))
  # Creating a folder that contain 5 subset with 100 times with a total of 500
  fold5 = caret::createMultiFolds(y = SL$Grain_Yield, k = 5, times = 1)

  for(i in 1:length(fold5)){
    index = fold5[[i]] # the index of the sample for training set
    #subset the phenotypic data
    train.data <- droplevels(SL %>%
                               filter(row_number() %in% index)) # subset the training set
    dim(train.data)
    test.data <- droplevels(SL %>%
                              filter(!row_number() %in% index)) # subset the testing set
    dim(test.data)

    test.data[,TraitN] = NA # change the grain yield of the training set to NA value

    mod_dat = rbind(train.data, test.data) # combine the the data set for analysis
    head(mod_dat)

  for(Trait in TraitN){
    eval(parse(text = paste("ans1 <- mmer(",Trait,"~1,
               random=~vsr(geno, Gu=A1) +
                 vsr(rowf) +
                 vsr(colf) +
                 spl2Da(row,col),
               rcov=~units,
               data= mod_dat)")))


    ### Extract the effects of the factor variables in the model
    prd = predict(object = ans1, classify = "geno") # predict the genotype mean value
    pr = prd$pvals # subset the estimate/ predicted value
    ind = which(pr$geno %in% test.data$geno) # indentify the genotypes in the test set from the predicted value
    pred_levl = as.character(pr$geno[ind]) # The levels of genotypes in the predicted table
    geno_test.set = test.data$geno[pred_levl] # Subset the genotypes from the test data
    length(levels(geno_test.set)) # the number of the genotyeps in test.set

    pr1 = pr %>% filter(geno %in% pred_levl) # filtering the predicted value based on the geno_commun
    ts1 = SL %>% filter(geno %in% pred_levl) # filtering the observed value of the test.set

    ts2 = ts1 %>% group_by(geno) %>%
      summarise_at(.vars = paste(Trait), .funs = mean) # Estimate teh mean of the test.set for the trait of interest
    ts2 = as.data.frame(ts2)
    # Estimate the RMSE and MAE vlaues between the predicted and observed value
    trt = paste(Trait)
    RM = cbind(location = paste(env),Trait = paste(Trait),RMSE = round(RMSE(pr1$predicted.value, ts2[,trt]),3),
               MAE = round(MAE(pr1$predicted.value, ts2[,trt]),3))
    # Estimate the predictability among the estiamted and predicted value
    preid = cbind(location = paste(env),Trait = paste(Trait),
                  predictability = round(cor(pr1$predicted.value,
                                             ts2[,trt],
                                             use = "pairwise.complete.obs"),3)) # estimate the correlation of the obseved and the predicted value

    Accu = rbind(Accu,RM) # store the measure of accuracy in tibble
    pr_y = rbind(pr_y,preid) # store model predictability in tibble

}
}
}
```

#Write the output in excel
```{r}
out = list(Accuracy = Accu, Predictability = pr_y)
out = cbind(Accu,pr_y)
library(openxlsx)
write.xlsx(x = out, file = "Analysis_out_with_marker_spatial_sommer.xlsx")
mean(as.numeric(pr_y$predictability), na.rm = T)


```

